### PR TITLE
feat: add AtomicData.from_structure() constructor to AtomicData for pymatgen support

### PR DESCRIPTION
**Pull request by atulcthakur**
# ALCHEMI Toolkit Pull Request

## Description

 Adds `from_structure()` to convert pymatgen `Structure` and `Molecule` objects directly to `AtomicData`, similar to the existing `from_atoms()` for ASE. Handles periodic, mixed-periodic, and non-periodic systems. Optional fields (energies, forces, stress, etc.) are only set when present in the structure's properties/site_properties - default to `None`.

 Also adds `pymatgen` as a project dependency.

## Type of Change

- [ ] Bug fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Performance improvement
- [ ] Documentation update
- [ ] Refactoring (no functional changes)
- [ ] CI/CD or infrastructure change

## Changes Made

  - `nvalchemi/data/atomic_data.py` — new `from_structure()` classmethod supporting both `Structure` and `Molecule`
  - `test/data/test_atomic_data.py` — 8 tests covering bulk, molecule, mixed pbc, optional fields, charge handling, and equivalence with `from_atoms`
  - `pyproject.toml` — added pymatgen dependency

## Testing

- [x] Unit tests pass locally (`make pytest`)
- [x] Linting passes (`make lint`)
- [x] New tests added for new functionality meets coverage expectations?

## Checklist

- [x] I have read and understand the [Contributing Guidelines](../CONTRIBUTING.md)
- [ ] I have updated the [CHANGELOG.md](../CHANGELOG.md)
- [x] I have performed a self-review of my code
- [ ] I have added docstrings to new functions/classes
- [ ] I have updated the documentation (if applicable)

## Additional Notes

> [!TIP]
> This repository uses Greptile, an AI code review service, to help conduct
> pull request reviews. We encourage contributors to read and consider suggestions
> made by Greptile, but note that human maintainers will provide the necessary
> reviews for merging: Greptile's comments are **not** a qualitative judgement
> of your code, nor is it an indication that the PR will be accepted/rejected.
> We encourage the use of emoji reactions to Greptile comments, depending on
> their usefulness and accuracy.

### CODE REVIEW

File: nvalchemi/data/atomic_data.py [hunk 1]
```diff
@@ -853,6 +853,177 @@ def from_atoms(
...
+            if not any(pbc_tuple):
+                pbc = None
+                cell = None
+            else:
+                cell = torch.as_tensor(
+                    structure.lattice.matrix.copy().reshape(1, 3, 3),
+                    device=device,
+                    dtype=dtype,
+                )
+                pbc = torch.as_tensor(pbc_tuple, device=device).reshape(1, 3)
```
[the remark this answers is not in the document]
Comment (atulcthakur) [reply nvalchemi/data/atomic_data.py:935]: This isn't needed for pymatgen. pymatgen will raise during construction if cell is invalid.

File: nvalchemi/data/atomic_data.py [hunk 2]
```diff
@@ -853,6 +853,177 @@ def from_atoms(
...
+
+        if node_charges is not None and node_charges.ndim == 1:
+            node_charges.unsqueeze_(-1)
+
+        # Derive graph-level charge — only set if explicitly provided
+        # Derive graph-level charge
+        if structure._charge is not None:
+            charge = torch.as_tensor(
+                [[int(structure.charge)]], device=device, dtype=dtype
+            )
```
[the remark this answers is not in the document]
Comment (atulcthakur) [reply nvalchemi/data/atomic_data.py]: This is an informed choice. Otherwise, we'd need to set default charge to 0.0, which leaves no way to identify if the 0.0 is coming from default or actually being set by the user.

File: nvalchemi/data/atomic_data.py [hunk 3]
```diff
@@ -853,6 +853,177 @@ def from_atoms(
...
+            node_attrs = to_one_hot(
+                indices.unsqueeze(-1),
+                num_classes=len(z_table),
+            ).to(dtype)
+
+        masses = torch.tensor(
+            [float(site.specie.atomic_mass) for site in structure],
+            device=device,
+            dtype=dtype,
+        )
```
[the remark this answers is not in the document]
Comment (atulcthakur) [reply nvalchemi/data/atomic_data.py:1049]: Ok, although I doubt the deprecation.

File: pyproject.toml [hunk 4]
```diff
@@ -13,6 +13,7 @@ requires-python = ">=3.11,<3.14"
 license = {file="LICENSE"}
 dependencies = [
     "ase>=3.27.0",
+    "pymatgen",
```
[the remark this answers is not in the document]
Comment (atulcthakur) [reply pyproject.toml]: Ok.

**Comment by atulcthakur** [comment]
/ok to test [1bd7c9c](https://github.com/NVIDIA/nvalchemi-toolkit/pull/36/commits/1bd7c9ca098698658464954480549c9d35c04459)

File: pyproject.toml [hunk 5]
```diff
@@ -13,6 +13,7 @@ requires-python = ">=3.11,<3.14"
 license = {file="LICENSE"}
 dependencies = [
     "ase>=3.27.0",
+    "pymatgen>=2025.10.7",
```
Comment (laserkelvin) [inline pyproject.toml]: Add this as an optional dependency. We should refactor `ase` to be optional as well, but we'll leave that to another PR

File: test/data/test_atomic_data.py [hunk 6]
```diff
@@ -710,3 +710,144 @@ def test_fully_periodic_works(self):
         assert data.pbc is not None
         assert data.cell.shape == (1, 3, 3)
         assert data.pbc.shape == (1, 3)
+
+
+# -----------------------------------------------------------------------------
+# from_structure: pymatgen Structure/Molecule conversion
+# -----------------------------------------------------------------------------
+class TestFromStructure:
```
Comment (laserkelvin) [inline test/data/test_atomic_data.py:732]: After moving `pymatgen` to optional, you'll want to add `pytest.importorskip('pymatgen')` to each test case

File: nvalchemi/data/atomic_data.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (laserkelvin) [reply nvalchemi/data/atomic_data.py:1049]: Could you just confirm it doesn't bring up the warning? We'll downvote if it's not true

File: nvalchemi/data/atomic_data.py [hunk 7]
```diff
@@ -853,6 +853,196 @@ def from_atoms(
...
+            virials_key,
+            dipole_key,
+        }
+        local_info: dict[str, torch.Tensor] = {}
+        for key, value in structure.properties.items():
+            if key in _consumed_props_keys:
+                continue
+            if isinstance(value, (np.ndarray, list)):
+                local_info[key] = torch.as_tensor(value, device=device, dtype=dtype)
+            elif isinstance(
```
Comment (laserkelvin) [inline nvalchemi/data/atomic_data.py:1004]: There's no third case?

Either way, please guard with `else: raise ...`

File: nvalchemi/data/atomic_data.py [hunk 8]
```diff
@@ -853,6 +853,196 @@ def from_atoms(
...
+            ) and not isinstance(value, (bool, np.bool_)):
+                local_info[key] = torch.as_tensor([value], device=device, dtype=dtype)
+
+        # Derive graph-level charge
+        if structure._charge is not None:
+            _charge = structure.charge
+            if abs(_charge - round(_charge)) >= 1e-2:
+                raise ValueError(f"Structure charge must be an integer, got {_charge}")
+            charge = torch.as_tensor(
+                [[int(round(_charge))]], device=device, dtype=dtype
```
Comment (laserkelvin) [inline nvalchemi/data/atomic_data.py:1021]: Why do you need to round and cast to int? It's probably set as a float but close to integers already right?

File: nvalchemi/data/atomic_data.py [hunk 9]
```diff
@@ -853,6 +853,196 @@ def from_atoms(
...
+                pbc = torch.as_tensor(pbc_tuple, device=device).reshape(1, 3)
+        else:
+            pbc = None
+            cell = None
+
+        # Extract optional fields from properties (system-level)
+        # and site_properties (per-atom).
+        raw_energy = structure.properties.get(energy_key)
+        energy = (
+            torch.as_tensor(raw_energy, device=device, dtype=dtype).reshape(1, 1)
```
Comment (laserkelvin) [inline nvalchemi/data/atomic_data.py]: For consistency, could you just do `[[raw_energy]]`?

File: pyproject.toml (hunk 5 above)
Comment (atulcthakur) [reply pyproject.toml]: pmg is now optional.

File: nvalchemi/data/atomic_data.py (hunk 7 above)
Comment (atulcthakur) [reply nvalchemi/data/atomic_data.py:1004]: I guarded. There's no guard in-place for `from_atoms`.

File: nvalchemi/data/atomic_data.py (hunk 8 above)
Comment (atulcthakur) [reply nvalchemi/data/atomic_data.py:1021]: It's a bit defensive. `int(1.999)` would give 1, but `int(round(1.999))` would give 2 correctly.

Comment (laserkelvin) [reply nvalchemi/data/atomic_data.py:1021]: I'm kind of inclined to not round, in case users pass something deliberately

**Review by laserkelvin — APPROVED** [review]
Will approve for now, but please create an issue to track the fact that `from_structure` does not guard against environments without `pymatgen` installed.

Ideally we want it to raise an exception and tell people to install `pymatgen`, but probably needs to be dressed up a bit more nicely
